pkg/controller/sparkapplication/controller.go

```diff
@@ -175,71 +176,83 @@ func (c *Controller) Stop() {
+	// The spec has changed. This is currently best effort as we can potentially miss updates
+	// and end up in an inconsistent state.
+	if !reflect.DeepEqual(oldApp.Spec, newApp.Spec) {
+		// Force-set the application status to Invalidating which handles clean-up and application re-run.
+		if err := c.updateSparkApplicationStatusWithRetries(newApp, func(status *v1alpha1.SparkApplicationStatus) {
+			status.AppState.State = v1alpha1.InvalidatingState
+		}); err != nil {
+			c.recorder.Eventf(
+				newApp,
+				apiv1.EventTypeWarning,
+				"SparkApplicationUpdateFailed",
+				"Failed to process update for SparkApplication %s, Error: %s.",
+				newApp.Name,
+				err)
+			return
+		} else {
+			c.recorder.Eventf(
+				newApp,
+				apiv1.EventTypeNormal,
+				"SparkApplicationUpdateWarning",
+				"SparkApplication %s was successfully updated and enqueued for submission. In some cases, SparkApplication updates can be missed."+
+					" A more reliable way of achieving this is to delete and re-create the SparkApplication.",
+				newApp.Name)
 		}
-		return
 	}
 
-	if oldApp.Status.DriverInfo.PodName != "" {
-		// Clear the application ID if the driver pod of the old application is to be deleted. This is important as
-		// otherwise deleting the driver pod of the old application if it's still running will result in a driver state
-		// update by the sparkPodMonitor with the driver pod phase set to PodFailed. This may lead to a restart of the
-		// application if it'c subject to a restart because of the failure state update of the driver pod. Clearing the
-		// application ID causes the state update regarding the old driver pod to be ignored in
-		// processSingleDriverStateUpdate because of mismatched application IDs and as a consequence no restart to be
-		// triggered. This prevents the application to be submitted twice: one from the restart and one from below.
-		c.updateSparkApplicationStatusWithRetries(newApp, func(status *v1alpha1.SparkApplicationStatus) {
-			status.AppID = ""
-		})
-	}
-
-	// Delete the driver pod and UI service of the old application. Note that deleting the driver pod kills the
-	// application if it is still running. Skip submitting the new application if cleanup for the old application
-	// failed to avoid potentially running both the old and new applications at the same time.
-	if err := c.deleteDriverAndUIService(oldApp, true); err != nil {
-		glog.Errorf("failed to delete the old driver pod and UI service for SparkApplication %s: %v",
-			oldApp.Name, err)
-		return
-	}
-
-	glog.Infof("SparkApplication %s was updated, enqueueing it for submission", newApp.Name)
+	glog.V(2).Infof("Spark Application %s enqueued for Processing.", newApp.GetName())
 	c.enqueue(newApp)
-	c.recorder.Eventf(
-		newApp,
-		apiv1.EventTypeNormal,
-		"SparkApplicationUpdated",
-		"SparkApplication %s was updated and enqueued for submission",
-		newApp.Name)
 }
 
 func (c *Controller) onDelete(obj interface{}) {
-	c.dequeue(obj)
```

Comment:
This should be kept. When an app gets deleted, it should be forgotten by the queue.

Reply:
Makes sense.